File: alljoyn/services/notification/cpp/src/Notification.cc
/******************************************************************************
 * Copyright (c) 2013-2014, AllSeen Alliance. All rights reserved.
 *
 *    Permission to use, copy, modify, and/or distribute this software for any
 *    purpose with or without fee is hereby granted, provided that the above
 *    copyright notice and this permission notice appear in all copies.
 *
 *    THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
 *    WITH REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF
 *    MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
 *    ANY SPECIAL, DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
 *    WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN
 *    ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
 *    OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
 ******************************************************************************/

#include <alljoyn/notification/Notification.h>
#include <alljoyn/notification/NotificationService.h>
#include "Transport.h"
#include "NotificationConstants.h"
#include <alljoyn/services_common/AsyncTaskQueue.h>
#include <alljoyn/notification/NotificationAsyncTaskEvents.h>
#include <qcc/Log.h>
#include <qcc/String.h>
#include <alljoyn/notification/LogModule.h>

using namespace ajn;
using namespace services;
using namespace qcc;

NotificationAsyncTaskEvents Notification::m_NotificationAsyncTaskEvents;
AsyncTaskQueue Notification::m_AsyncTaskQueue(&Notification::m_NotificationAsyncTaskEvents);

Notification::Notification(int32_t messageId,
                           NotificationMessageType messageType, const char* deviceId, const char* deviceName,
                           const char* appId, const char* appName, const char* sender,
                           std::map<qcc::String, qcc::String> const& customAttributes,
                           std::vector<NotificationText> const& notificationText,
                           const char* richIconUrl, std::vector<RichAudioUrl> const&  richAudioUrl,
                           const char* richIconObjectPath, const char* richAudioObjectPath,
                           const char* controlPanelServiceObjectPath, const char* originalSender) :
    m_MessageId(messageId), m_Sender(0), m_MessageType(messageType), m_DeviceId(0),
    m_DeviceName(0), m_AppId(0), m_AppName(0), m_CustomAttributes(customAttributes),
    m_Text(notificationText), m_RichIconUrl(0), m_RichAudioUrl(richAudioUrl), m_RichIconObjectPath(0),
    m_RichAudioObjectPath(0), m_ControlPanelServiceObjectPath(0), m_OriginalSender(0)
{
    setDeviceId(deviceId);
    setDeviceName(deviceName);
    setAppId(appId);
    setAppName(appName);
    setSender(sender);
    setRichIconUrl(richIconUrl);
    setRichIconObjectPath(richIconObjectPath);
    setRichAudioObjectPath(richAudioObjectPath);
    setControlPanelServiceObjectPath(controlPanelServiceObjectPath);
    setOriginalSender(originalSender);
}

Notification::Notification(NotificationMessageType messageType, std::vector<NotificationText> const& notificationText) :
    m_MessageId(-1), m_Sender(0),  m_MessageType(messageType), m_DeviceId(0), m_DeviceName(0), m_AppId(0), m_AppName(0), m_Text(notificationText), m_RichIconUrl(0),  m_RichIconObjectPath(0),
    m_RichAudioObjectPath(0), m_ControlPanelServiceObjectPath(0), m_OriginalSender(0)

{
}

Notification::Notification(const Notification& notification)
{
    m_Sender = NULL;
    m_DeviceId = NULL;
    m_DeviceName = NULL;
    m_AppId = NULL;
    m_AppName = NULL;
    m_RichIconUrl = NULL;
    m_RichIconObjectPath = NULL;
    m_RichAudioObjectPath = NULL;
    m_ControlPanelServiceObjectPath = NULL;
    m_OriginalSender = NULL;

    setDeviceId(notification.getDeviceId());
    setDeviceName(notification.getDeviceName());
    setAppId(notification.getAppId());
    setAppName(notification.getAppName());
    setSender(notification.getSenderBusName());
    setRichIconUrl(notification.getRichIconUrl());
    setRichIconObjectPath(notification.getRichIconObjectPath());
    setRichAudioObjectPath(notification.getRichAudioObjectPath());
    setControlPanelServiceObjectPath(notification.getControlPanelServiceObjectPath());
    setOriginalSender(notification.getOriginalSender());

    m_CustomAttributes = notification.m_CustomAttributes;
    m_MessageId = notification.m_MessageId;
    m_MessageType = notification.m_MessageType;
    m_Text = notification.m_Text;
    m_RichAudioUrl = notification.m_RichAudioUrl;
}

Notification::~Notification()
{
    delete m_Sender;
    m_Sender = NULL;

    delete m_DeviceId;
    m_DeviceId = NULL;

    delete m_DeviceName;
    m_DeviceName = NULL;

    delete m_AppId;
    m_AppId = NULL;

    delete m_AppName;
    m_AppName = NULL;

    delete m_RichIconUrl;
    m_RichIconUrl = NULL;

    delete m_RichIconObjectPath;
    m_RichIconObjectPath = NULL;

    delete m_RichAudioObjectPath;
    m_RichAudioObjectPath = NULL;

    delete m_ControlPanelServiceObjectPath;
    m_ControlPanelServiceObjectPath = NULL;

    delete m_OriginalSender;
    m_OriginalSender = NULL;
}

const uint16_t Notification::getVersion() const
{
    return NotificationService::getVersion();
}

const char* Notification::getDeviceId() const
{
    if (m_DeviceId == NULL) {
        return NULL;
    }
    return m_DeviceId->c_str();
}

const char* Notification::getDeviceName() const
{
    if (m_DeviceName == NULL) {
        return NULL;
    }
    return m_DeviceName->c_str();
}

const char* Notification::getAppId() const
{
    if (m_AppId == NULL) {
        return NULL;
    }
    return m_AppId->c_str();
}

const char* Notification::getAppName() const
{
    if (m_AppName == NULL) {
        return NULL;
    }
    return m_AppName->c_str();
}

const std::map<qcc::String, qcc::String>& Notification::getCustomAttributes() const
{
    return m_CustomAttributes;
}

const int32_t Notification::getMessageId() const
{
    return m_MessageId;
}

const std::vector<NotificationText>& Notification::getText() const
{
    return m_Text;
}

const char* Notification::getSenderBusName() const
{
    if (m_Sender == NULL) {
        return NULL;
    }
    return m_Sender->c_str();
}

const NotificationMessageType Notification::getMessageType() const
{
    return m_MessageType;
}

const char* Notification::getRichIconUrl() const
{
    if (m_RichIconUrl == NULL) {
        return NULL;
    }
    return m_RichIconUrl->c_str();
}

const char* Notification::getRichIconObjectPath() const
{
    if (m_RichIconObjectPath == NULL) {
        return NULL;
    }
    return m_RichIconObjectPath->c_str();
}

const char* Notification::getRichAudioObjectPath() const
{
    if (m_RichAudioObjectPath == NULL) {
        return NULL;
    }
    return m_RichAudioObjectPath->c_str();
}

const std::vector<RichAudioUrl>& Notification::getRichAudioUrl() const
{
    return m_RichAudioUrl;
}

const char* Notification::getControlPanelServiceObjectPath() const
{
    if (m_ControlPanelServiceObjectPath == NULL) {
        return NULL;
    }
    return m_ControlPanelServiceObjectPath->c_str();
}

const char* Notification::getOriginalSender() const
{
    if (m_OriginalSender == NULL) {
        return NULL;
    }
    return m_OriginalSender->c_str();
}

void Notification::setAppId(const char* appId) {
    if (appId == NULL) {
        delete m_AppId;
        m_AppId = NULL;
        return;
    }

    if (!m_AppId) {
        m_AppId = new qcc::String(appId);
    } else {
        m_AppId->assign(appId);
    }
}

void Notification::setAppName(const char* appName) {
    if (appName == NULL) {
        delete m_AppName;
        m_AppName = NULL;
        return;
    }

    if (!m_AppName) {
        m_AppName = new qcc::String(appName);
    } else {
        m_AppName->assign(appName);
    }
}

void Notification::setControlPanelServiceObjectPath(
    const char* controlPanelServiceObjectPath) {
    if (controlPanelServiceObjectPath == NULL) {
        delete m_ControlPanelServiceObjectPath;
        m_ControlPanelServiceObjectPath = NULL;
        return;
    }

    if (!m_ControlPanelServiceObjectPath) {
        m_ControlPanelServiceObjectPath = new qcc::String(controlPanelServiceObjectPath);
    } else {
        m_ControlPanelServiceObjectPath->assign(controlPanelServiceObjectPath);
    }
}

void Notification::setCustomAttributes(
    const std::map<qcc::String, qcc::String>& customAttributes) {
    m_CustomAttributes = customAttributes;
}

void Notification::setDeviceId(const char* deviceId) {
    if (deviceId == NULL) {
        delete m_DeviceId;
        m_DeviceId = NULL;
        return;
    }

    if (!m_DeviceId) {
        m_DeviceId = new qcc::String(deviceId);
    } else {
        m_DeviceId->assign(deviceId);
    }
}

void Notification::setDeviceName(const char* deviceName) {
    if (deviceName == NULL) {
        delete m_DeviceName;
        m_DeviceName = NULL;
        return;
    }

    if (!m_DeviceName) {
        m_DeviceName = new qcc::String(deviceName);
    } else {
        m_DeviceName->assign(deviceName);
    }
}

void Notification::setOriginalSender(const char* originalSender) {
    if (originalSender == NULL) {
        delete m_OriginalSender;
        m_OriginalSender = NULL;
        return;
    }

    if (!m_OriginalSender) {
        m_OriginalSender = new qcc::String(originalSender);
    } else {
        m_OriginalSender->assign(originalSender);
    }
}

void Notification::setMessageId(int32_t messageId) {
    m_MessageId = messageId;
}

void Notification::setRichAudioUrl(
    const std::vector<RichAudioUrl>& richAudioUrl) {
    m_RichAudioUrl = richAudioUrl;
}

void Notification::setRichIconUrl(const char* richIconUrl) {
    if (richIconUrl == NULL) {
        delete m_RichIconUrl;
        m_RichIconUrl = NULL;
        return;
    }

    if (!m_RichIconUrl) {
        m_RichIconUrl = new qcc::String(richIconUrl);
    } else {
        m_RichIconUrl->assign(richIconUrl);
    }
}

void Notification::setRichIconObjectPath(const char* richIconObjectPath) {
    if (richIconObjectPath == NULL) {
        delete m_RichIconObjectPath;
        m_RichIconObjectPath = NULL;
        return;
    }

    if (!m_RichIconObjectPath) {
        m_RichIconObjectPath = new qcc::String(richIconObjectPath);
    } else {
        m_RichIconObjectPath->assign(richIconObjectPath);
    }
}

void Notification::setRichAudioObjectPath(const char* richAudioObjectPath) {
    if (richAudioObjectPath == NULL) {
        delete m_RichAudioObjectPath;
        m_RichAudioObjectPath = NULL;
        return;
    }

    if (!m_RichAudioObjectPath) {
        m_RichAudioObjectPath = new qcc::String(richAudioObjectPath);
    } else {
        m_RichAudioObjectPath->assign(richAudioObjectPath);
    }
}

void Notification::setSender(const char* sender) {
    if (sender == NULL) {
        delete m_Sender;
        m_Sender = NULL;
        return;
    }

    if (!m_Sender) {
        m_Sender = new qcc::String(sender);
    } else {
        m_Sender->assign(sender);
    }
}

QStatus Notification::dismiss()
{
    QCC_DbgPrintf(("Notification::dismiss() called OriginalSender:%s MessageId:%d AppId:%s", getOriginalSender(), getMessageId(), getAppId()));

    NotificationMsg* notificationMsg = new NotificationMsg(getOriginalSender(), getMessageId(), getAppId());
    m_AsyncTaskQueue.Enqueue(notificationMsg);

    return ER_OK;
}

File: alljoyn/services/notification/java/NotificationService/src/org/alljoyn/ns/Notification.java
/******************************************************************************
 * Copyright (c) 2014, AllSeen Alliance. All rights reserved.
 *
 *    Permission to use, copy, modify, and/or distribute this software for any
 *    purpose with or without fee is hereby granted, provided that the above
 *    copyright notice and this permission notice appear in all copies.
 *
 *    THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
 *    WITH REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF
 *    MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
 *    ANY SPECIAL, DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
 *    WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN
 *    ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
 *    OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
 ******************************************************************************/

package org.alljoyn.ns;

import java.util.List;
import java.util.Map;
import java.util.UUID;

import org.alljoyn.ns.transport.consumer.NotificationFeedback;

/**
 * The Notification that is sent over the network
 */
public class Notification {
	          
	/**
	 * version
	 */
	private int version;		
	
	/**
	 * Notification message Id
	 */
	private int messageId;				
	
	/**
	 * Notification message type
	 */
	private NotificationMessageType messageType;	
	
	/**
	 * App name
	 */
	private String appName;		
	
	/**
	 * App Id
	 */
	private UUID appId;	
	
	/**
	 * Device name
	 */
	private String deviceName;		
	
	/**
	 * Device Id
	 */
	private String deviceId;	
			
	/**
	 * Sender Id - name of the bus that the {@link Notification} was received from 
	 */
	private String sender;
	
	/**
	 * The sender Id of the Bus that originally sent the {@link Notification} message.
	 * If the {@link Notification} was received from a Super Agent, the sender field of the object is of the Super Agent, 
	 * and this field is the value of the {@link Notification} producer that has originally sent the message.    
	 */
	private String origSender;
	
	/**
	 * Map of customized array variable
	 */
	private Map<String, String> customAttributes;		
	
	/**
	 * list of NotificationText objects
	 */
	private List<NotificationText> text;			
	
	/**
	 * Rich notification icon URL
	 */
	private String richIconUrl;
	
	/**
	 * Rich notification audio url
	 */
	private List<RichAudioUrl> richAudioUrl;
	
	/**
	 * Rich notification icon object path
	 */
	private String richIconObjPath;
	
	/**
	 * Rich notification audio object path
	 */
	private String richAudioObjPath;
	
	/**
	 * Control Panel Service object path
	 */
	private String responseObjectPath;
	
	/**
	 * Constructor
	 * @param messageType The type of the sent or received Notification
	 * @param text The list of text messages to be sent or received
	 */
	public Notification(NotificationMessageType messageType, List<NotificationText> text) throws NotificationServiceException {
		setMessageType(messageType);
		setText(text);
	}//Notification

	/**
	 * @return The protocol version
	 */
	public int getVersion() {
		return version;
	}

	/**
	 * @return Notification message type
	 */
	public NotificationMessageType getMessageType() {
		return messageType;
	}

	/**
	 * @return Returns the application name
	 */
	public String getAppName() {
		return appName;
	}

	/**
	 * @return The application identifier
	 */
	public UUID getAppId() {
		return appId;
	}
	
	/**
	 * The device name that has sent the {@link Notification}
	 */
	public String getDeviceName() {
		return deviceName;
	}

	/**
	 * The device id that has sent the {@link Notification}
	 */
	public String getDeviceId() {
		return deviceId;
	}

	/**
	 * Sender Id - name of the bus that the {@link Notification} was received from
	 */
	public String getSenderBusName() {
		return sender;
	}
	
	/**
	 * The sender Id of the Bus that originally sent the {@link Notification} message.
	 * If the {@link Notification} was received from a Super Agent, the sender field of the object is of the Super Agent, 
	 * and this field is the value of the {@link Notification} producer that has originally sent the message.
	 * @return Bus name of the original {@link Notification} sender
	 */
	public String getOriginalSenderBusName() {
		return origSender;
	}
	
	/**
	 * Map of customized array variable
	 */
	public Map<String, String> getCustomAttributes() {
		return customAttributes;
	}

	/**
	 * List of {@link NotificationText} objects
	 */
	public List<NotificationText> getText() {
		return text;
	}

	/**
	 * {@link Notification} message id
	 */
	public int getMessageId() {
		return messageId;
	}

	/**
	 * @see org.alljoyn.ns.Notification#getRichIconUrl()
	 */
	public String getRichIconUrl() {
		return richIconUrl;
	}
	
	/**
	 * Rich notification icon URL
	 */
	public List<RichAudioUrl> getRichAudioUrl() {
		return richAudioUrl;
	}	
	
	/**
	 * Rich notification icon object path
	 */
	public String getRichIconObjPath() {
		return richIconObjPath;
	}
	
	/**
	 * Rich notification audio object path
	 */
	public String getRichAudioObjPath() {
		return richAudioObjPath;
	}
	
	/**
	 * The object identifier to be used to invoke methods of the sender.
	 * Usually used for Notification with action.
	 */
	public String getResponseObjectPath() {
		return responseObjectPath;
	}
	
	/**
	 * Set notification message type
	 * @param messageType
	 * @throws NotificationServiceException Is thrown if the message type is NULL
	 */
	public void setMessageType(NotificationMessageType messageType) throws NotificationServiceException {
		if ( messageType == null ) {
			throw new NotificationServiceException("MessageType must be set");
		}
		this.messageType = messageType;
	}//setMessageType
	
	/**
	 * Set customAttributes map
	 * @param customAttributes
	 */
	public void setCustomAttributes(Map<String, String> customAttributes) {
		this.customAttributes = customAttributes;
	}
	
	/**
	 * Set list of NotificationText
	 * @param text Is thrown if the text is NULL or is empty
	 */
	public void setText(List<NotificationText> text) throws NotificationServiceException {
		if ( text == null || text.size() == 0 ) {
			throw new NotificationServiceException("The text argument must be set and may not be an empty list");
		}
		this.text = text;
	}//setText

	/**
	 * Set the icon url for displaying rich content
	 * @param richIconUrl 
	 */
	public void setRichIconUrl(String richIconUrl) {
		this.richIconUrl = richIconUrl;
	}
	
	/**
	 * Set audio url to send rich content
	 * @param richAudioUrl - A list of url per language
	 */
	public void setRichAudioUrl(List<RichAudioUrl> richAudioUrl) {
		this.richAudioUrl = richAudioUrl;
	}
	
	/**
	 * Set the icon Object Path for displaying rich content
	 * @param richIconObjPath 
	 */
	public void setRichIconObjPath(String richIconObjPath) {
		this.richIconObjPath = richIconObjPath;
	}
	
	/**
	 * Set audio Object Path to send rich content
	 * @param richAudioObjPath - An object path for the rich audio object
	 */
	public void setRichAudioObjPath(String richAudioObjPath) {
		this.richAudioObjPath = richAudioObjPath;
	}

	/**
	 * Set Control Panel Service Object Path
	 * @param responseObjectPath
	 */
	public void setResponseObjectPath(String responseObjectPath) {
		this.responseObjectPath = responseObjectPath;
	}

	
	//=======================================//
	
	/**
	 * When the notification message is dismissed, it is first of all deleted by the Notification Producer
	 * and then a dismiss signal is sent to update other Notification Consumers that the {@link Notification} 
	 * has been dismissed   
	 */
	public void dismiss() {
		try {
			new NotificationFeedback(this).dismiss();
		} catch (NotificationServiceException nse) {
			System.out.println("Failed to call the dismiss method, Error: '" + nse.getMessage() + "'");
		}
	}//dismiss

	/**
	 * @see java.lang.Object#toString()
	 */
	@Override
	public String toString() {
		
		StringBuilder sb = new StringBuilder("Notification: ");
		sb.append("Id: '").append(messageId).append("'")
		.append(", MsgType: '").append(messageType).append("'")
		.append(", DeviceId: '").append(deviceId).append("'")
		.append(", DeviceName: '").append(deviceName).append("'")
		.append(", Version: '").append(version).append("'")
		.append(", Sender: '").append(sender).append("'")
		.append(", OrigSender: '").append(origSender).append("'")
		.append(", AppId: '").append(appId).append("'")
		.append(", AppName: '").append(appName).append("'")
		.append(", CustomAttributes: '").append(customAttributes).append("'")
		.append(", TextMessage: '").append(text).append("'");
		
		if ( richIconUrl != null && richIconUrl.length() > 0 ) {
			sb.append(", RichIconURL: '").append(richIconUrl).append("'");
		}
		
		if ( richIconObjPath != null && richIconObjPath.length() > 0 ) {
			sb.append(", RichIconObjPath: '").append(richIconObjPath).append("'");
		}
		
		if ( richAudioUrl != null && richAudioUrl.size() > 0 ) {
			sb.append(", RichAudioUrl: '").append(richAudioUrl).append("'");
		}
		
		if ( richAudioObjPath != null && richAudioObjPath.length() > 0 ) {
			sb.append(", RichAudioObjPath: '").append(richAudioObjPath).append("'");
		}
		
		if ( responseObjectPath != null && responseObjectPath.length() > 0 ) {
			sb.append(", ResponseObjPath: '").append(responseObjectPath).append("'");
		}
		
		return sb.toString();
	}//toString
	
	//=======================================//
	


	/**
	 * Set message signature version
	 * @param version
	 */
	void setVersion(int version) {
		this.version = version;
	}
		
	/**
	 * Set message Id
	 * @param messageId
	 */
	void setMessageId(int messageId) {
		this.messageId = messageId;
	}
	
	/**
	 * Set app name
	 * @param appName
	 */
	void setAppName(String appName) {
		this.appName = appName;
	}
	
	/**
	 * Set app Id
	 * @param appId
	 */
	void setAppId(UUID appId) {
		this.appId = appId;
	}
	
	/**
	 * Set device name
	 * @param appName
	 */
	void setDeviceName(String deviceName) {
		this.deviceName = deviceName;
	}
	
	/**
	 * Set device Id
	 * @param deviceId
	 */
	void setDeviceId(String deviceId) {
		this.deviceId = deviceId;
	}
	
	/**
	 * Set sender Id
	 * @param sender
	 */
	void setSender(String sender) {
		this.sender = sender;
	}
	
	/**
	 * Set the original sender name
	 * @param origSender
	 */
	void setOrigSender(String origSender) {
		this.origSender = origSender;
	}
	
}//NotificationImpl
